ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp

```diff
@@ -21,44 +22,43 @@ namespace ngraph
+            void mean(const T* arg, T* out, const Shape& in_shape, const AxisSet& reduction_axes)
             {
-                auto out_shape = reduce(in_shape, reduction_axes, keep_dims);
-                CoordinateTransform output_transform(out_shape);
-                std::vector<T> cs(shape_size(out_shape));
+                constexpr bool dont_keep_dims_in_output = false;
+                const auto out_shape = reduce(in_shape, reduction_axes, dont_keep_dims_in_output);
+                std::vector<T> cs(shape_size(out_shape), 0);
+                std::fill(out, out + shape_size(out_shape), 0);
 
-                for (const Coordinate& output_coord : output_transform)
-                {
-                    out[output_transform.index(output_coord)] = 0;
-                    cs[output_transform.index(output_coord)] = 0;
-                }
+                const auto in_strides = row_major_strides(in_shape);
+                const auto out_strides = row_major_strides(out_shape);
 
-                CoordinateTransform input_transform(in_shape);
+                CoordinateTransformBasic input_transform(in_shape);
                 std::map<size_t, int> index_to_count_map;
 
                 for (const Coordinate& input_coord : input_transform)
                 {
-                    Coordinate output_coord = reduce(input_coord, reduction_axes, keep_dims);
+                    const Coordinate output_coord =
+                        reduce(input_coord, reduction_axes, dont_keep_dims_in_output);
+
+                    const size_t in_idx = std::inner_product(
+                        input_coord.begin(), input_coord.end(), in_strides.begin(), 0);
+                    const size_t out_idx = std::inner_product(
+                        output_coord.begin(), output_coord.end(), out_strides.begin(), 0);
 
-                    T x = arg[input_transform.index(input_coord)];
-                    T& z = out[output_transform.index(output_coord)];
-                    auto index = output_transform.index(output_coord);
-                    if (index_to_count_map.find(index) == index_to_count_map.end())
+                    T x = arg[in_idx];
+                    T& z = out[out_idx];
+                    if (index_to_count_map.find(out_idx) == index_to_count_map.end())
                     {
-                        index_to_count_map[index] = 1;
+                        index_to_count_map[out_idx] = 1;
                     }
                     else
                     {
-                        index_to_count_map[index]++;
+                        index_to_count_map[out_idx]++;
                     }
 
                     if (is_finite(x) && is_finite(z))
                     {
-                        T& c = cs[output_transform.index(output_coord)];
+                        T& c = cs[out_idx];
                         T t = z + (x - c);
                         c = (t - z) - (x - c);
                         z = t;
```

Comment:
I know that this PR is rather about removing `index()` call so I approve that. 
But for me this piece of code is not easily understandable, 
to compare that `else` block is a simple sum.

Reply:
I think while we are revising reference implementations we need to solve all issues.
At least I really don't understand the variables names (what does it mean? What is `z`, `x` and etc.?)

Reply:
Looks like: https://en.wikipedia.org/wiki/Kahan_summation_algorithm

Reply:
I have added helper function with description. I hope this helps to understand.